### PR TITLE
A combination of mine and crisbeto@'s recent work on diagnostics in the compiler

### PR DESCRIPTION
**Pull request by alxhub**
This PR merges together several independent PRs: #34457, #34333, #34061 

These were all interdependent and so this PR resolves the conflicts between them. Each constituent PR has already been reviewed & approved.

### CODE REVIEW

**Comment by alxhub** [comment]
[Presubmit](http://test/OCL:286105671:BASE:286105699:1576637250442:407b77b0)
[Ivy Presubmit](http://test/OCL:286105671:BASE:286105699:1576637250469:5b07c5b)

**Review by crisbeto — APPROVED** [review]

File: packages/compiler-cli/test/ngtsc/scope_spec.ts [hunk 1]
```diff
@@ -301,6 +301,87 @@ runInEachFileSystem(() => {
...
+          export class DepModule {}
+
+          @NgModule({
+            declarations: [TestCmp],
+            imports: [DepModule],
+          })
+          export class Module {}
+        `);
+
+        for (const diag of env.driveDiagnostics()) {
```
Comment (JoostK) [inline packages/compiler-cli/test/ngtsc/scope_spec.ts]: nit: I would personally assign the results of `env.driveDiagnostics()` to a variable, to help debugging if needed.

**Review by JoostK — APPROVED** [review]
I have only reviewed the first 3 commits by Alex, assuming that the other ones have already been approved.

**Comment by alxhub** [comment]
[Updated Presubmit](http://test/OCL:286250310:BASE:286250357:1576703398831:7349709)
[Updated Ivy Presubmit](https://test.corp.google.com/ui#id=OCL:286250310:BASE:286250357:1576703398873:358ae8c1)

**Comment by kara** [comment]
caretaker note to me: Alex is an owner of fw-ngcc, so Joost's LGTM is sufficient.

**Pull request closed without merging**